OSDOCS-4626: Add embedding MicroShift

fzdarsky opened the pull request:
Signed-off-by: Frank A. Zdarsky <fzdarsky@redhat.com>

ShaunaDiaz commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+    " /usr/share/osbuild-composer/repositories/rhel-87.json | sudo tee /etc/osbuild-composer/repositories/rhel-87.json
+----
+
+Verify the entries have been added. They should look like the following:
```
Is there a command to run to verify?

ShaunaDiaz replied:
We'd replace "They should look like the following:" with the .Example styling. Take a look at microshift_networking/modules/microshift-ki-cni-iptables-deleted.adoc

ShaunaDiaz commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+$ dnf download microshift-release-info-{ocp-version}.0-1.el{rhel-major}
+----
+
+. Exctract the release info by entering the following command:
```
. Extract the release info by running the following command:

fzdarsky replied:
Updated.

ShaunaDiaz commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+$ dnf download microshift-release-info-{ocp-version}.0-1.el{rhel-major}
+----
+
+. Exctract the release info by entering the following command:
```
"running" the command is Docs style

fzdarsky replied:
Updated.

ShaunaDiaz commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+source = "<image_pullspec_with_tag_or_digest>"
+----
+
+To embed MicroShift's container images, you need to know the exact list of
```
You need to have the exact list of container image references used by the {product-title} version to embed {product-title} 's container images. Use the `microshift-release-info` RPM package
matching the version of the `microshift` RPM in your blueprint.

.Procedure

fzdarsky replied:
Updated.

ShaunaDiaz commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+
+...
+----
+
```
needs id

ShaunaDiaz commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+* You must have downloaded an installation link:https://console.redhat.com/openshift/install/pull-secret[pull secret] from the Red Hat Hybrid Cloud Console and make this available in `/etc/crio/openshift-pull-secret` with root user-only read/write permissions.
+* You must configure the firewall with the mandatory firewall rules.
+
+If you are using any of the methods using a kickstart file, you can provide these prerequsites, for example, by adding the following snippet to your kickstart file:
```
Sounds like another procedure: Using a kickstart file

ShaunaDiaz commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+methods described in the https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/composing_installing_and_managing_rhel_for_edge_images/index[{op-system-ostree} documentation], you must configure the system to meet the
+following prerequisites:
+
+* You must provide a logical volume manager (LVM) volume group (VG) with sufficient capacity for the persistent volumes (PVs) of your workload.
```
Do we have procedures for the first two items?
Can xref our firewall procedure

dhellmann replied:
LVM is a standard RHEL thing, I don't think we want to redocument it here. If we can easily link to some RHEL doc, that's fine, but I think we have to expect a certain level of understanding of operating system concepts in these docs so if there's no easy link available I think we can get along fine without it.

ShaunaDiaz commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+// Module included in the following assemblies:
```
Start all modules with microshift-
For example, microshift-embedding-into-rpm-ostree-image.doc

ShaunaDiaz reviewed (commented):
Some comments here about style and structure. Not exhaustive. We'll need to rework things further into Docs style once content is settled, but I wanted to provide some feedback since you did the work. Thank you!

dhellmann commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+
+If your use case requires running the `oc` CLI tool locally on your MicroShift
+machine, further add the `openshift-clients` package.
+If your use case requires using MicroShift's "https://[auto-deployed manifests]"
```
Is that a missing link?

dhellmann commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+        \"name\": \"fast-datapath-for-rhel-8-$arch-rpms\",
+        \"baseurl\": \"https://cdn.redhat.com/content/dist/layered/rhel8/$arch/fast-datapath/os\"}
+    )]
+    " /usr/share/osbuild-composer/repositories/rhel-87.json | sudo tee /etc/osbuild-composer/repositories/rhel-87.json
```
Does this tee command actually work? I would expect the shell to invoke tee so it has a stdout handle to pass to jq, which would overwrite the file before jq can read the file to manipulate the content.

fzdarsky replied:
Note source and destination files are different (different dirs).

dhellmann replied:
Ah, I totally missed that.

dhellmann reviewed (commented):
These instructions look OK. I'm not thrilled with the jq commands, but don't see a better option right now, so we may just want to consider whether we want to create another tool or wrapper script for the next release.

dhellmann commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+enabled = ["microshift"]
+----
+
+If your use case requires running the `oc` CLI tool locally on your MicroShift
```
They would add a packages stanza for the new package, using the same style as above. We could give that example separately but it seems like that would complicate the doc here. Maybe we should just always include it in the example above instead of introducing the conditional?

fzdarsky replied:
I fear if we include it, users may interpret this as recommendation / requirement. Also, it’d no longer be a *minimal* example.

dhellmann replied:
OK, I'm looking for ways to keep the doc simple. I think @ShaunaDiaz point is that these 2 sentences are not enough by themselves. So do we want a completely separate example procedure for the remote git repo case (I could see how an end-to-end example of that would be helpful, but also a lot to write)? If we want it, I think I'd put off trying to write that until 4.13 opens.

ShaunaDiaz commented:
@fzdarsky Do you want to continue working on this yourself today, or do you want me to take over your PR?

fzdarsky commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+[id="embedding-microshift-rpm-ostree-image_{context}"]
+= Creating a {op-system-ostree} image embedding MicroShift
+== Adding MicroShift repositories to Image Builder
+Image Builder's `osbuild-composer` backend does not inherit the build
```
Done.

fzdarsky commented on microshift_install/microshift-embed-rpm-ostree.adoc:
```diff
@@ -1,23 +1,25 @@
+:context: microshift-embed-rpm-ostree
 toc::[]
 
 You can embed MicroShift into a {op-system-ostree-first} {op-system-version} image.
```
Done.

fzdarsky commented on microshift_install/microshift-embed-rpm-ostree.adoc:
```diff
@@ -1,23 +1,25 @@
 :_content-type: ASSEMBLY
-[id="microshift-install-rhel-for-edge"]
+[id="microshift-embed-rpm-ostree"]
 = Embedding MicroShift into a {op-system-ostree} image
```
Updated.

fzdarsky commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+Image Builder's `osbuild-composer` backend does not inherit the build
+host's system repos in `/etc/yum.repos.d/`. Instead, it uses its own set of
+official repositories defined in the `/usr/share/osbuild-composer/repositories`
+directory. When adding MicroShift's RPM repositories to Image Builder, you
```
Updated.

fzdarsky commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+feature with manifests stored on a remote git repository, further add the `git`
+package.
+
+You can further customize your image, e.g. to add users, firewall rules, or
```
Updated.

fzdarsky commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/composing_installing_and_managing_rhel_for_edge_images/composing-a-rhel-for-edge-image-using-image-builder-command-line_composing-installing-managing-rhel-for-edge-images#image-customizations_composing-a-rhel-for-edge-image-using-image-builder-command-line[supported image customizations].
+
+=== Adding the MicroShift container images
+Image Builder supports building rpm-ostree images with embedded container
```
Updated.

fzdarsky commented:
@ShaunaDiaz I've refactored the PR so each section is in a separate module, rewrote the "adding repo" section to use a different procedure, and updated the other sections to use .Example and "Additional Resources". Hope I understood correctly what you asked. Can you please have another look?

fzdarsky commented on modules/embedding-microshift-into-rpm-ostree-image.adoc:
```diff
@@ -0,0 +1,212 @@
+
+== Adding MicroShift to an Image Builder blueprint
+=== Adding the MicroShift service
+To build an {op-system-ostree} image with MicroShift, simply add the MicroShift
```
Updated.

ShaunaDiaz reviewed (commented):
Per today's meeting, I opened https://github.com/openshift/openshift-docs/pull/54388 to replace this one and to finalize for 4.12. I couldn't close your PR, but you can go ahead and close it, referencing the new one. Thank you for working so hard on this!

ggiguash commented on modules/microshift-adding-containers-to-blueprint.adoc:
```diff
@@ -0,0 +1,60 @@
+You need to have the exact list of container image references used by the
+{product-title} version to embed {product-title} 's container images.
+Use the `microshift-release-info` RPM package matching the version of the
+`microshift` RPM in your blueprint.
```
This paragraph kind of repeats the previous one, so they might better be merged.

fzdarsky replied:
Good catch. Yes, meant to delete the "To embed..." one with the second one suggested by Shauna.

ggiguash commented on modules/microshift-adding-containers-to-blueprint.adoc:
```diff
@@ -0,0 +1,60 @@
+[source,terminal]
+----
+$ yum download microshift-release-info-4.12.0-1.el8
+----
```
I would automate current version extraction using the following command:
```
VERSION=$(rpm -q --queryformat='%{VERSION}' microshift)
```

fzdarsky replied:
That's a good idea. It'd probably have to be `'%{VERSION}-%{RELEASE}` to pin the exact package, though, or? (Haven't tested.)

ggiguash commented on modules/microshift-adding-containers-to-blueprint.adoc:
```diff
@@ -0,0 +1,60 @@
+
+[[containers]]
+source = "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:82cfef91557f9a70cff5a90accba45841a37524e9b93f98a97b20f6b2b69e5db"
+
```
Let's separate the command from the file contents?

fzdarsky replied:
That's the output of the command. IDK whether there's a doc style guide for how to deal with output, just copied this from somewhere else.

ggiguash commented on modules/microshift-adding-repos-to-image-builder.adoc:
```diff
@@ -0,0 +1,74 @@
++
+[source,terminal]
+----
+$ sudo subscription-manager repos --enable rhocp-4.12-for-rhel-8-$(uname -i)-rpms --enable fast-datapath-for-rhel-8-$(uname -i)-rpms
```
```suggestion
$ sudo subscription-manager repos \
   --enable rhocp-4.12-for-rhel-8-$(uname -i)-rpms \
   --enable fast-datapath-for-rhel-8-$(uname -i)-rpms
```

ggiguash commented on modules/microshift-adding-repos-to-image-builder.adoc:
```diff
@@ -0,0 +1,74 @@
+[source,terminal]
+----
+$ sudo reposync --arch=$(uname -i) --arch=noarch --gpgcheck \
+    --download-path /var/repos/microshift-local \
```
`/var/repos` is not a standard location. I'm not sure we should be creating top-level dirs in /var.
Suggesting to use `/var/cache/microshift/repos`, also used by osbuild, etc.

ggiguash replied:
Also, let's set repo base path in a variable to make it easier for the users to relocate if needed.
```
REPO_PATH=/var/cache/microshift/repos
```

fzdarsky replied:
`/var/repo` is the location used in the RH KB article (https://access.redhat.com/solutions/23016) referenced from the RHEL docs. I didn't use `/var/cache/$app` as this is normally for data that'd be recreated by the corresponding $app, but I don't think the exact location doesn't matter, as long as it's not in a user dir and accessible by the `_osbuild` user.

fzdarsky commented:
Moving this content into a new PR #54388.
